### PR TITLE
Added sync indicator

### LINKED ISSUE #2606: Create indicator to show content is syncing
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

This project will add a syncing indicator to the User icon in order to show that the app is still refreshing with the most recent information.

You'll use a pre-existing Onyx key to hook into the API loading state (relevant API information [here](https://github.com/Expensify/Expensify.cash/pull/2524/files)) and display a syncing indicator on the User icon as per the **Expected Results** below.

## Expected Result:

While the app is syncing, the small green availability indicator (`AvatarWithIndicator`) should grow in size and a small syncing icon should appear:

When the loading is complete, the icon should return to a regular green dot.

## Actual Result:
Currently, there is no UI change when the app is syncing.

## Action Performed:

1. Install and log into the Expensify.cash mobile app
2. Close the app and wait long enough that new information should need to sync
3. Open the app to observe the behaviour that is occurring while syncing

## Platform:
Where is this issue occurring?

- [x] Web
- [x]  iOS
- [x]  Android
- [x]  Desktop App
- [x]  Mobile Web

**Version Number:** v.1.011
**Logs:** N/A
**Notes/Photos/Videos:** 
_**Additional context from original issue:** when I open up the mobile app it shows that I'm connected and shows all of the content downloaded locally which is great, but I always know I need to wait a couple seconds because it's in the process of synchronizing down new content. It would be nice if it shows some kind of feedback indicating that it is in the process of actively reconnecting, so I know when I can stop waiting. I'm thinking some kind of animation around the Green Dot, or even just making the Green Dot blink or something. At this point the app is functional and you can click around, but its content is potentially out of date. I wouldn't want to introduce additional seconds into the startup process while waiting to complete synchronization._

**Expensify/Expensify Issue URL:** https://github.com/Expensify/Expensify/issues/159767

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

### PR DESCRIPTION
**Pull request by parasharrajat**
Please review.

### Details

Added sync indicator which is shown when `ONYXKEYS.IS_LOADING_AFTER_RECONNECT` is `true`

### Fixed Issues

Fixes #2606

#### Web/Desktop
1. Open the JavaScript Dev console and go to the network tab.
1. Switch the app to offline (see demo video below with screenshots)
1. Switch the app back online. Verify that the indicator is shown.

#### Web/Desktop
1. Open the JavaScript Dev console and go to the network tab.
1. Switch the app to offline (see demo video below with screenshots)
1. Switch the app back online. Verify that the indicator is shown.

#### iOS/Android/mWeb
1. Put the app on background on mobile, then going back to foreground indicator should be shown
2. Put on Airplane mode, then turn it off, the indicator will be visible.

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web

https://user-images.githubusercontent.com/24370807/118722955-67976d80-b84a-11eb-830d-e5fc22777fb2.mp4

#### Mobile Web

https://user-images.githubusercontent.com/24370807/118724799-afb78f80-b84c-11eb-8c52-8f0ae80a977e.mp4

#### Desktop

https://user-images.githubusercontent.com/24370807/118723751-67e43880-b84b-11eb-813e-971c65731792.mp4

#### iOS

https://user-images.githubusercontent.com/24370807/119175386-61420500-ba87-11eb-9c3a-491d615daa0b.mp4
#### Android

https://user-images.githubusercontent.com/24370807/118723013-7a11a700-b84a-11eb-9712-a1bfb4e2f3ef.mp4

### CODE REVIEW

File: src/components/AvatarWithIndicator.js
Comment (shawnborton) [inline src/components/AvatarWithIndicator.js:121]: Could you try making the icon 12x12? This way there is a little bit of margin around the icon inside of the 16x16 green circle. 

image

Comment (parasharrajat) [reply src/components/AvatarWithIndicator.js:121]: @shawnborton But indicator size is 12x12 including 2px transparent border around. So Green circle is 8x8.

Comment (shawnborton) [reply src/components/AvatarWithIndicator.js:121]: We want the indicator to become larger when we are showing that content is syncing. So the green online indicator should grow in size and then show the spinning icon inside of it. Let me know if that makes sense - happy to make a prototype as well.

Comment (parasharrajat) [reply src/components/AvatarWithIndicator.js:121]: OK after growing, it should be 16px. makes sense.

Comment (parasharrajat) [reply src/components/AvatarWithIndicator.js:121]: @shawnborton It looks a little big. Don't you think so?
image

Comment (parasharrajat) [reply src/components/AvatarWithIndicator.js:121]: Or this is looking fine, let me know I will update the PR.

File: src/components/AvatarWithIndicator.js
Comment (TomatoToaster) [inline src/components/AvatarWithIndicator.js:134]: NAB: For my understanding, you got rid of the memo here because you're animating this right? I'm guessing when you were testing this it wouldn't wouldn't animate while this was wrapping.

File: src/pages/home/sidebar/SidebarLinks.js
Comment (TomatoToaster) [inline src/pages/home/sidebar/SidebarLinks.js:75]: This is picky but would you mind changing this comment to match the other comments something like:
```suggestion
    // Whether we are syncing app data
```

File: src/components/AvatarWithIndicator.js
Comment (parasharrajat) [reply src/components/AvatarWithIndicator.js:134]: No I used PureComponent and we don't need it now.

File: src/components/AvatarWithIndicator.js
Comment (TomatoToaster) [inline src/components/AvatarWithIndicator.js:22]: Also picky but do you think it would be more clear to change this to:
```suggestion
    // Whether we show the sync indicator
```

**Review by TomatoToaster — CHANGES_REQUESTED** [review]

**Review by TomatoToaster — DISMISSED** [review]
LGTM - I'm might get someone to double check me. But I'd also wait for Shawn to get back to you on the icon size.

**Comment by roryabraham** [comment]
Added myself as a review to help buddy-check @TomatoToaster. @parasharrajat Will review tomorrow 👍

File: src/components/AvatarWithIndicator.js
Comment (parasharrajat) [reply src/components/AvatarWithIndicator.js:121]: @shawnborton any thoughts?

**Comment by roryabraham** [comment]
Hey @parasharrajat can you please include screenshots for the other platforms? Also, I edited the tests + qa steps to clarify them and add platform-specific instructions. Also, do you have any recommendations for testing this on an iOS simulator/Android emulator?

**Comment by parasharrajat** [comment]
Let me add screenshots for all devices.

File: src/components/AvatarWithIndicator.js
Comment (roryabraham) [inline src/components/AvatarWithIndicator.js:93]: We've tried to take a pretty hard stance against inline styles. In this scenario, we would usually make a function in `styles.js` called `getAvatarIndicatorStyles` or something and dynamically generate these styles in there. You can see how I handed transforms in a dynamic style function in `src/styles/getTooltipSyles.js`. We haven't really decided whether dynamic style generators should be in their own module or directly in `styles.js`, so it's up to you! Since it's probably just one function, I would probably put it directly in `styles.js`

**Comment by shawnborton** [comment]
Apologies @parasharrajat as I was out of office yesterday. I'll review this soon!

File: src/components/AvatarWithIndicator.js
Comment (roryabraham) [reply src/components/AvatarWithIndicator.js:121]: I agree it looks big. I think ideally the size of the indicator wouldn't need to change – we just show the spinning icon in the normal-size indicator if it's loading.

File: src/components/AvatarWithIndicator.js
Comment (roryabraham) [inline src/components/AvatarWithIndicator.js:48]: Since this is a `PureComponent`, do you need to implement `shouldComponentUpdate` to make sure it re-renders if `this.props.isSyncing` changes? If it tests fine without it then no need, just thinking out loud.

**Review by roryabraham — CHANGES_REQUESTED** [review]

File: src/components/AvatarWithIndicator.js
Comment (shawnborton) [reply src/components/AvatarWithIndicator.js:121]: We can try a smaller size, which would be a green circle of 12x12, with the icon inside of it at 8x8. It would look like this:
image

image

**Comment by roryabraham** [comment]
@parasharrajat During my testing on iOS I was able to successfully test this the first time the app loads. However, repeating the same steps I am unable to get it to work again. Could this be related to needing `shouldComponentUpdate`?

FWIW I see three ways to test this locally on iOS:

1. (only Expensify employees) Set `USE_NGROK=true` in your `.env` to point the app at your VM, and then turn ngrok off and back on to simulate the loss and regain of a network connection.
2. (iOS Simulator) Follow [these instructions](https://nshipster.com/network-link-conditioner/) to set up Network Link Conditioner. 
3. Run the app on a physical device and turn airplane mode on and off.

**Comment by roryabraham** [comment]
Looking at my ngrok logs it seems like the problem is like this:

1. I turn off ngrok, so the app registers that it's offline and starts logging many "possible unhandled promise rejection" warnings.
2. I turn ngrok back on, but it doesn't notice and start clearing the network queue again. So the loading indicator doesn't show.

**Comment by parasharrajat** [comment]
@roryabraham Sorry. I just found something is off but yet to find where exactly. Animation does not seem to Loop well. Please hold your testing until I push a fix to that.

 I see successful state change without shouldComponentUpdate as In absence of this by default shallow matching will be done for props and thus it will without adding custom `shouldComponentUpdate`.

**Comment by parasharrajat** [comment]
@shawnborton Does this look right?

https://user-images.githubusercontent.com/24370807/118708387-4bd79b80-b839-11eb-8a81-8196db870c14.mp4

**Comment by shawnborton** [comment]
This looks great Rajat! Could you just _slightly_ speed up the speed of the icon rotating? Otherwise I think it's good to go.

**Comment by parasharrajat** [comment]
Ready for review.

File: src/components/AvatarWithIndicator.js
Comment (shawnborton) [inline src/components/AvatarWithIndicator.js:137]: Just to make sure I am following, when the indicator is in syncing mode, the small white icon that spins is shown at 6x6?

Comment (parasharrajat) [reply src/components/AvatarWithIndicator.js:137]: I am not setting the width & height for the indicator. Instead, I use scaling which is Smooth. So when we scale the green dot, all its children including the icon will scale. 8px icon was looking big when scaling so I have adjusted it and found 6px is looking fine.

Scaling using fixed px is not smooth as the indicator is absolutely placed thus when we increase width & height, It changes location and does not look good. Oppose to what scaling keeps the center and scale uniformly.

Comment (shawnborton) [reply src/components/AvatarWithIndicator.js:137]: Ah, that makes sense to me and sounds good!

**Review by TomatoToaster — DISMISSED** [review]

**Comment by roryabraham** [comment]
@parasharrajat When testing this on iOS I was able to get it in a state where it is "stuck" syncing.

**Comment by parasharrajat** [comment]
Ok. Let me check it with Network Link Conditioner.

**Comment by parasharrajat** [comment]
@roryabraham Not sure how to turn off the WIfi with NLC. I would definitely need that tip but I tried 100% loss profile and On bringing the app on foreground from background, I also saw that syncing never stops~~(takes like 5 mins). I checked the logs it seems that `ONYXKEYS.IS_LOADING_AFTER_RECONNECT` state is not updating in the Onyx thus stop is never called afterward. This logic completely relies on `ONYXKEYS.IS_LOADING_AFTER_RECONNECT` key. But eventually, it was stopped.

**Comment by roryabraham** [comment]
> Not sure how to turn off the WIfi with NLC. I would definitely need that tip but I tried 100% loss profile

Yeah, this is what I meant by "turn off Wifi". Sorry it that wasn't clear.

> This logic completely relies on ONYXKEYS.IS_LOADING_AFTER_RECONNECT key.

I think that since this PR introduces a visual indication of the state of that key, we should not merge it until the key is being successfully updated when we expect it to.

**Comment by parasharrajat** [comment]
> I think that since this PR introduces a visual indication of the state of that key, we should not merge it until the key is being successfully updated when we expect it to.

Agree, let me see what happens with the `ONYXKEYS.IS_LOADING_AFTER_RECONNECT` Key in that state.

**Comment by parasharrajat** [comment]
Strange, now it seems to work fine. I checked the Triggering logic and we are using Promises.all So it is possible that some of the API calls hanged forever which is unlikely. It should not happen with real devices even if there is a very poor connection. The only thing I can think of at the moment is Timeout the calls.

**Comment by roryabraham** [comment]
@parasharrajat Sorry, I still can't seem to get this working on iOS. 😕 It seems to spin forever, regardless of the state of the network connection. Can you provide a screencast of it working on iOS? The code _seems_ fine, so it could be something weird with ngrok, or my VM or something. I don't want to block you on this forever, but I also can't approve this without seeing it work on all platforms. If you can provide a screencast of this working on iOS I might be willing to just merge this and see if it works on staging.

**Comment by parasharrajat** [comment]
Sure @roryabraham. I will share the Screencast.

**Comment by roryabraham** [comment]
@TomatoToaster Did this work for you?

**Comment by parasharrajat** [comment]
@roryabraham Here is a recording. Its little Blurry due to resolution but functionality can be seen well.

https://user-images.githubusercontent.com/24370807/119175386-61420500-ba87-11eb-9c3a-491d615daa0b.mp4

**Review by roryabraham — APPROVED** [review]
Okay, looks good to me. Thanks for providing screen recordings on all platforms.

**Comment by roryabraham** [comment]
@shawnborton Code looks good to me. I'll let you to a final design review, then if everything looks good feel free to merge once all the checks pass.

**Review by shawnborton — APPROVED** [review]

**Comment by shawnborton** [comment]
Looks good to me design-wise!

**Pull request merged**
